### PR TITLE
Watchtime

### LINKED ISSUE #5412: Watch time
### Description 
We need a way to display watch time per users for screenings. These stats will be saved and always displayed so we shall keep it. 

### Specs
- unity: min
- saved and loaded from firebase

### Acceptance criteria
- [x] First, validate count strategy 
Then implement on event form
- [x] New screening stat column with watch time per user
- [x] Average watch time on a screening

image

### PR DESCRIPTION
**Pull request by pldespaigne**
fix #5412

### CODE REVIEW

File: apps/festival/festival/src/app/marketplace/event/session/session.component.ts [hunk 1]
```diff
@@ -49,10 +51,27 @@ export class SessionComponent implements OnInit, OnDestroy {
 
   private countdownId: number = undefined;
 
+  private watchTime: number; // watch-time in secondes
+  private watchTimeIntervalId: number;
+  private invitationId: string;
+  private _playingState: 'play' | 'pause' = 'pause';
+  public get playingState() { return this._playingState; }
+  public set playingState(value: 'play' | 'pause') {
```
Comment (GrandSchtroumpf) [inline apps/festival/festival/src/app/marketplace/event/session/session.component.ts:59]: Do you expect other state ? If not use a boolean

File: apps/festival/festival/src/app/marketplace/event/session/session.component.ts [hunk 2]
```diff
@@ -77,6 +96,30 @@ export class SessionComponent implements OnInit, OnDestroy {
         if (!!(event.meta as Screening).titleId) {
           const movie = await this.movieService.getValue(event.meta.titleId as string);
           this.screeningFileRef = movie.promotional.videos?.screener;
+
+          // if user is not a screening owner we need to track the watch time
+          if (event.ownerOrgId !== this.authQuery.orgId) {
```
Comment (GrandSchtroumpf) [inline apps/festival/festival/src/app/marketplace/event/session/session.component.ts:101]: I would rather have a stream with filters & observables on playing state with an `interval(60 * 1000)` stream. Like that we can subscribe/unsubscribe easily to it, and have one source of update

File: libs/event/src/lib/components/analytics/analytics.component.ts [hunk 3]
```diff
@@ -20,20 +22,49 @@ export class EventAnalyticsComponent implements OnInit {
...
   constructor(
     private service: EventService,
+    private invitationQuery: InvitationQuery,
     private cdr: ChangeDetectorRef,
   ) { }
 
   async ngOnInit() {
-    const transformEvent = (event: EventAnalytics) => ({ ...event, name: `${event.firstName} ${event.lastName}` });
-    const analytics = await this.service.queryAnalytics(this.eventId);
+    const transformEvent = (event: EventAnalytics) => {
```
Comment (GrandSchtroumpf) [inline libs/event/src/lib/components/analytics/analytics.component.ts:39]: I've got a hard time understanding the code below.

File: libs/invitation/src/lib/+state/invitation.firestore.ts [hunk 4]
```diff
@@ -27,6 +27,9 @@ export interface InvitationBase<D> extends PublicInvitation{
    */
   eventId?: string;
   message?: string;
+
+  /** Watch time in secondes, only used for 'screening' events */
```
Comment (RemcoSimonides) [inline libs/invitation/src/lib/+state/invitation.firestore.ts:31]: seconds*

Comment (pldespaigne) [reply libs/invitation/src/lib/+state/invitation.firestore.ts:31]: 🥐  🇫🇷 🥖

File: apps/festival/festival/src/app/marketplace/event/session/session.component.ts [hunk 5]
```diff
@@ -77,6 +87,30 @@ export class SessionComponent implements OnInit, OnDestroy {
...
+            this.watchTimeInterval = interval(1000).pipe(
+              filter(() => !!this.isPlaying),
+              scan(watchTime => watchTime + 1, invitation.watchTime ?? 0),
+              finalizeWithValue(watchTime => {
+                if (watchTime !== undefined) this.invitationService.update(this.invitationId, { watchTime });
+              }),
+              filter(watchTime => watchTime % 60 === 0),
+            ).subscribe(watchTime => {
+              this.invitationService.update(this.invitationId, { watchTime });
+            });
```
Comment (RemcoSimonides) [inline apps/festival/festival/src/app/marketplace/event/session/session.component.ts:112]: Curious if I understand this correctly;
The watchtime is updated every 60 seconds or when finalized, right?

**Review by RemcoSimonides — APPROVED** [review]
Well done

File: apps/festival/festival/src/app/marketplace/event/session/session.component.ts (hunk 5 above)
Comment (pldespaigne) [reply apps/festival/festival/src/app/marketplace/event/session/session.component.ts:112]: exactly !

**Pull request merged**